lib/engine/game/g_1870/game.rb

```diff
@@ -932,13 +937,13 @@ def legal_tile_rotation?(_entity, hex, tile)
             end
         end
 
-        def upgrades_to?(from, to, _special = false)
-          return to.name == '170' if from.color == :green && P_HEXES.include?(from.hex.name)
+        def upgrades_to?(from, to, special = false)
+          return false if (to.name == '171K') ^ (from.color == :brown && from.hex.name == 'B11')
```

Comment:
Do we need xor here?

Reply:
It was a decision between that or repeating all the huge lines with slightly different connectives. I opted for xor because I thought it was slightly more informative of the intent (iff) and more compact.

For example, that line would become
```
          return false if (to.name != '171K') && (from.color == :brown && from.hex.name == 'B11')
          return false if (to.name == '171K') && (from.color != :brown || from.hex.name != 'B11')
```

Or some variation of this, which makes it easier to follow from examples, but makes the high-level idea more opaque.

Maybe I could add a comment like _`return if p ^ q` means something like 'assert p <=> q'`

Reply:
I'll have to wait for @tobymao's input since we haven't used XORs in the code before.

The other changes look fine.

Reply:
There's no hurry :)

It's good we try to keep this as readable as possible, because this code is always tricky (I can't tell you how many iterations I went through until I figured out a clean way to catch all the cases without rewriting the main method...)

Reply:
what are you trying to do here?

Reply:
In the case of this line I'm saying that the tile 171K can only be placed on B11 and that if B11 is brown, it can only be upgraded to the tile 171K.

Now that I'm reading the lines I've wrote expanding the xor, they can indeed be simplified:
```
return false if to.name != '171K' && (from.color == :brown && from.hex.name == 'B11')
return false if to.name == '171K' && from.hex.name != 'B11'
```

I'll leave it however you prefer...

Reply:
This would normally be done with Labels, but the labelled cities use normal yellow and green tiles and I couldn't find a reasonable way to use them.

Other games with similar situation copy the entire original method (~20 lines) and make the necessary adjustments in place, but I don't think that's a very future-proof option

Reply:
if tile == '171.k'
  return (from.color == brown && from.hex )  || ...
end

Reply:
Well, this specific tile doesn't have any other restrictions, but for the brown P tiles, it still needs to check if the paths are respected. At this point we're just copying the entire `upgrades_to?` method multiple times, might as well just override it completely and not call `super`

Reply:
i don't have enough context to make the call, but it seems like there's some simplification you can do

Reply:
It would become:

```
        def upgrades_to?(from, to, special = false)
          # correct color progression?
          return false unless Engine::Tile::COLORS.index(to.color) == (Engine::Tile::COLORS.index(from.color) + 1)

          # honors pre-existing track?
          return false unless from.paths_are_subset_of?(to.paths)

          # correct label?
          return false if to.label && !from.label
          return false if to.label == 'K' && from.hex.name != 'B11'
          return false if to.label == 'L' && from.hex.name != 'C18'

          true
        end
```

Reply:
I just realised that setting `special = true` skips the labels check and all the other checks that it skips are irrelevant for 1870, so I'll submit a new version using that

Reply:
See if you like this version: #4329